Applications/Spire/Source/Ui/TagBox.cpp

```diff
@@ -0,0 +1,480 @@
+  layout->setContentsMargins({});
+  m_text_box = new TextBox(std::move(current_model));
+  set_style(*m_text_box, TEXT_BOX_STYLE(get_style(*m_text_box)));
+  m_text_box->installEventFilter(this);
+  m_list_view = new ListView(m_model,
+    std::bind_front(&TagBox::build_tag, this));
+  m_list_view->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+  set_style(*m_list_view, LIST_VIEW_STYLE(get_style(*m_list_view)));
+  m_list_view_style_connection = connect_style_signal(*m_list_view,
+    std::bind_front(&TagBox::on_list_view_style, this));
+  for(auto i = 0; i < m_list_view->get_list()->get_size(); ++i) {
+    set_style(*m_list_view->get_list_item(i), LIST_ITEM_STYLE());
+  }
+  m_list_view->get_list()->connect_operation_signal(
+    std::bind_front(&TagBox::on_operation, this));
+  m_list_view->connect_submit_signal(std::bind_front(&TagBox::on_submit, this));
+  m_list_view->setFocusPolicy(Qt::NoFocus);
+  m_list_view->installEventFilter(this);
+  m_list_view_container = new QWidget();
+  auto container_layout = new QHBoxLayout(m_list_view_container);
+  container_layout->setContentsMargins({});
+  container_layout->setSpacing(0);
+  container_layout->addWidget(m_list_view);
+  auto input_box = make_input_box(m_list_view_container);
+  layout->addWidget(input_box);
+  proxy_style(*this, *input_box);
+  set_style(*this, INPUT_BOX_STYLE(get_style(*input_box)));
+  m_style_connection = connect_style_signal(*this, [=] { on_style(); });
+  m_ellipses_item = m_list_view->get_list_item(get_list()->get_size());
+  m_ellipses_item->hide();
+  setFocusProxy(m_text_box);
+  setFocusPolicy(Qt::StrongFocus);
+  m_line_edit = m_text_box->findChild<QLineEdit*>();
+  connect(m_line_edit, &QLineEdit::cursorPositionChanged,
+    std::bind_front(&TagBox::reposition_list_view, this));
+  m_focus_connection = m_focus_observer.connect_state_signal(
+    std::bind_front(&TagBox::on_focus, this));
+}
+
+const std::shared_ptr<ListModel>& TagBox::get_list() const {
+  return m_model->m_source;
+}
+
+const std::shared_ptr<TextModel>& TagBox::get_current() const {
+  return m_text_box->get_current();
+}
+
+bool TagBox::is_read_only() const {
+  return m_text_box->is_read_only();
+}
+
+void TagBox::set_read_only(bool is_read_only) {
+  m_text_box->set_read_only(is_read_only);
+  update_uneditable();
+  if(is_read_only) {
+    match(*this, ReadOnly());
+    match(*m_list_view, ReadOnly());
+  } else {
+    unmatch(*this, ReadOnly());
+    unmatch(*m_list_view, ReadOnly());
+  }
+}
+
+connection TagBox::connect_delete_signal(
+    const DeleteSignal::slot_type& slot) const {
+  return m_delete_signal.connect(slot);
+}
+
+bool TagBox::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == m_text_box && event->type() == QEvent::KeyPress) {
+    auto& key_event = *static_cast<QKeyEvent*>(event);
+    switch(key_event.key()) {
+      case Qt::Key_Backspace:
+        if(m_line_edit->cursorPosition() == 0 && get_list()->get_size() > 0) {
```

Comment:
Use ```TextBox```'s selection model.

Reply:
Since I didn't find the selection model, I used the highlight model to get the cursor position.

Reply:
Yes yes, it's the highlight model not selection model. Good catch.